File: LBTAS.py
#!/usr/bin/env python3
"""
Leveson-Based Trade Assessment Scale (LBTAS)
============================================

A rigorous assessment methodology for digital commerce adapted from 
Nancy Leveson's Software Assessment Scale used in aircraft software development.

Copyright (C) 2024 Network Theory Applied Research Institute
Licensed under GNU Affero General Public License v3.0

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
"""

import json
import os
import sys
import argparse
from pathlib import Path
from typing import Dict, List, Optional, Union
from datetime import datetime

__version__ = "1.0.0"
__author__ = "Network Theory Applied Research Institute"
__license__ = "AGPL-3.0"

class LevesonRatingSystem:
    """
    Leveson-Based Trade Assessment Scale rating system.
    
    Provides a rigorous 6-point scale (-1 to +4) for evaluating
    digital commerce interactions based on Nancy Leveson's methodology.
    """
    
    DEFAULT_CATEGORIES = ['reliability', 'usability', 'performance', 'support']
    
    RATING_DESCRIPTIONS = {
        -1: "No Trust - User was harmed, exploited, or received a product or service with no discipline or malicious intent.",
        0: "Cynical Satisfaction - Interaction fulfills a basic promise requiring little to no discipline toward user satisfaction.",
        1: "Basic Promise - Interaction meets all articulated user demands, no more.",
        2: "Basic Satisfaction - Interaction meets socially acceptable standards exceeding articulated user demands.",
        3: "No Negative Consequences - Interaction designed to prevent loss, exceed basic quality.",
        4: "Delight - Interaction anticipates the evolution of user practices and concerns post-transaction."
    }
    
    def __init__(self, storage_file: Optional[str] = None, categories: Optional[List[str]] = None):
        """
        Initialize the LBTAS rating system.
        
        Args:
            storage_file: Optional path to JSON file for persistent storage
            categories: Custom rating categories (defaults to reliability, usability, performance, support)
        """
        self.categories = categories or self.DEFAULT_CATEGORIES.copy()
        self.storage_file = storage_file
        self.exchanges = {}
        
        if self.storage_file:
            self.load_from_file()
    
    def add_exchange(self, name: str) -> None:
        """Add a new exchange to the system."""
        if name in self.exchanges:
            raise ValueError(f"Exchange '{name}' already exists.")
        
        self.exchanges[name] = {
            category: [] for category in self.categories
        }
        self.exchanges[name]['_metadata'] = {
            'created': datetime.now().isoformat(),
            'total_ratings': 0
        }
        
        if self.storage_file:
            self.save_to_file()
    
    def add_rating(self, exchange_name: str, criterion: str, rating: int) -> None:
        """
        Add a rating programmatically (non-interactive).
        
        Args:
            exchange_name: Name of the exchange
            criterion: Rating category
            rating: Rating value (-1 to 4)
        """
        if exchange_name not in self.exchanges:
            raise ValueError(f"Exchange '{exchange_name}' does not exist.")
        
        if criterion not in self.categories:
            raise ValueError(f"Criterion '{criterion}' not in valid categories: {self.categories}")
        
        if not isinstance(rating, int) or rating < -1 or rating > 4:
            raise ValueError(f"Rating must be integer between -1 and 4, got {rating}")
        
        self.exchanges[exchange_name][criterion].append(rating)
        self.exchanges[exchange_name]['_metadata']['total_ratings'] += 1
        
        if self.storage_file:
            self.save_to_file()
    
    def get_rating(self, criterion: str) -> int:
        """Get a rating from the user for a specific criterion (interactive)."""
        print(f"\nRate {criterion.capitalize()}:")
        print("=" * 50)
        
        for rating, description in self.RATING_DESCRIPTIONS.items():
            print(f" {rating:2d}: {description}")
        
        print("=" * 50)
        
        while True:
            try:
                rating = int(input(f"Enter your rating for {criterion.capitalize()} (-1 to 4): "))
                if -1 <= rating <= 4:
                    return rating
                else:
                    print("Please enter a rating between -1 and 4.")
            except ValueError:
                print("Invalid input. Please enter a number between -1 and 4.")
            except KeyboardInterrupt:
                print("\nRating cancelled.")
                sys.exit(0)
    
    def rate_exchange(self, name: str) -> None:
        """Rate an exchange based on Leveson Software Assessment Scale (interactive)."""
        if name not in self.exchanges:
            raise ValueError(f"Exchange '{name}' does not exist.")
        
        print(f"\nRating '{name}' using Leveson-Based Trade Assessment Scale")
        print("=" * 60)
        
        for criterion in self.categories:
            rating = self.get_rating(criterion)
            self.exchanges[name][criterion].append(rating)
            self.exchanges[name]['_metadata']['total_ratings'] += 1
        
        if self.storage_file:
            self.save_to_file()
        
        print(f"\nRating completed for '{name}'!")
    
    def view_ratings(self, name: str) -> Dict[str, Optional[float]]:
        """View the average ratings for a specific exchange."""
        if name not in self.exchanges:
            raise ValueError(f"Exchange '{name}' does not exist.")
        
        ratings_summary = {}
        
        for criterion in self.categories:
            ratings = self.exchanges[name][criterion]
            if ratings:
                avg_rating = sum(ratings) / len(ratings)
                ratings_summary[criterion] = round(avg_rating, 2)
            else:
                ratings_summary[criterion] = None
        
        return ratings_summary
    
    def get_all_exchanges(self) -> List[str]:
        """Return a list of all exchanges."""
        return list(self.exchanges.keys())
    
    def generate_report(self) -> Dict:
        """Generate a comprehensive system report."""
        total_exchanges = len(self.exchanges)
        
        if total_exchanges == 0:
            return {
                'total_exchanges': 0,
                'system_average': None,
                'top_performers': [],
                'bottom_performers': [],
                'category_averages': {}
            }
        
        # Calculate system-wide averages
        all_ratings = []
        category_totals = {cat: [] for cat in self.categories}
        
        exchange_averages = {}
        
        for exchange_name, exchange_data in self.exchanges.items():
            exchange_ratings = []
            
            for category in self.categories:
                ratings = exchange_data[category]
                if ratings:
                    avg = sum(ratings) / len(ratings)
                    exchange_ratings.append(avg)
                    category_totals[category].extend(ratings)
                    all_ratings.extend(ratings)
            
            if exchange_ratings:
                exchange_averages[exchange_name] = sum(exchange_ratings) / len(exchange_ratings)
        
        # System average
        system_average = sum(all_ratings) / len(all_ratings) if all_ratings else None
        
        # Category averages
        category_averages = {}
        for category, ratings in category_totals.items():
            if ratings:
                category_averages[category] = sum(ratings) / len(ratings)
            else:
                category_averages[category] = None
        
        # Top and bottom performers
        sorted_exchanges = sorted(exchange_averages.items(), key=lambda x: x[1], reverse=True)
        top_performers = sorted_exchanges[:5]  # Top 5
        bottom_performers = sorted_exchanges[-5:]  # Bottom 5
        
        return {
            'total_exchanges': total_exchanges,
            'total_ratings': len(all_ratings),
            'system_average': round(system_average, 2) if system_average else None,
            'top_performers': top_performers,
            'bottom_performers': bottom_performers,
            'category_averages': {k: round(v, 2) if v else None for k, v in category_averages.items()},
            'generated_at': datetime.now().isoformat()
        }
    
    def save_to_file(self) -> None:
        """Save exchanges to JSON file."""
        if not self.storage_file:
            return
        
        try:
            with open(self.storage_file, 'w', encoding='utf-8') as f:
                json.dump(self.exchanges, f, indent=2, ensure_ascii=False)
        except Exception as e:
            print(f"Warning: Could not save to {self.storage_file}: {e}")
    
    def load_from_file(self) -> None:
        """Load exchanges from JSON file."""
        if not self.storage_file or not os.path.exists(self.storage_file):
            return
        
        try:
            with open(self.storage_file, 'r', encoding='utf-8') as f:
                data = json.load(f)
                # Validate loaded data structure
                for exchange_name, exchange_data in data.items():
                    if not isinstance(exchange_data, dict):
                        continue
                    # Ensure all categories exist
                    for category in self.categories:
                        if category not in exchange_data:
                            exchange_data[category] = []
                    # Ensure metadata exists
                    if '_metadata' not in exchange_data:
                        exchange_data['_metadata'] = {
                            'created': datetime.now().isoformat(),
                            'total_ratings': sum(len(ratings) for cat, ratings in exchange_data.items() if cat != '_metadata')
                        }
                
                self.exchanges = data
        except Exception as e:
            print(f"Warning: Could not load from {self.storage_file}: {e}")
    
    def export_to_csv(self, filename: str) -> None:
        """Export ratings to CSV format."""
        import csv
        
        with open(filename, 'w', newline='', encoding='utf-8') as csvfile:
            fieldnames = ['exchange', 'category', 'rating', 'timestamp']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
            
            writer.writeheader()
            for exchange_name, exchange_data in self.exchanges.items():
                for category in self.categories:
                    for rating in exchange_data[category]:
                        writer.writerow({
                            'exchange': exchange_name,
                            'category': category,
                            'rating': rating,
                            'timestamp': datetime.now().isoformat()
                        })

def main():
    """Command-line interface for LBTAS."""
    parser = argparse.ArgumentParser(
        description="Leveson-Based Trade Assessment Scale (LBTAS)",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  lbtas rate --exchange "MyService"
  lbtas add --exchange "MyService" --criterion reliability --rating 3
  lbtas view --exchange "MyService"
  lbtas report
  lbtas export --format csv --output ratings.csv
        """
    )
    
    parser.add_argument('--version', action='version', version=f'LBTAS {__version__}')
    parser.add_argument('--storage', default='lbtas_ratings.json', 
                       help='Storage file for ratings (default: lbtas_ratings.json)')
    
    subparsers = parser.add_subparsers(dest='command', help='Available commands')
    
    # Rate command (interactive)
    rate_parser = subparsers.add_parser('rate', help='Rate an exchange interactively')
    rate_parser.add_argument('--exchange', required=True, help='Exchange name to rate')
    
    # Add command (programmatic)
    add_parser = subparsers.add_parser('add', help='Add a rating programmatically')
    add_parser.add_argument('--exchange', required=True, help='Exchange name')
    add_parser.add_argument('--criterion', required=True, help='Rating criterion')
    add_parser.add_argument('--rating', type=int, required=True, help='Rating value (-1 to 4)')
    
    # View command
    view_parser = subparsers.add_parser('view', help='View ratings for an exchange')
    view_parser.add_argument('--exchange', required=True, help='Exchange name to view')
    
    # List command
    list_parser = subparsers.add_parser('list', help='List all exchanges')
    
    # Report command
    report_parser = subparsers.add_parser('report', help='Generate system report')
    
    # Export command
    export_parser = subparsers.add_parser('export', help='Export ratings')
    export_parser.add_argument('--format', choices=['json', 'csv'], default='json', help='Export format')
    export_parser.add_argument('--output', required=True, help='Output filename')
    
    args = parser.parse_args()
    
    if not args.command:
        parser.print_help()
        return
    
    # Initialize rating system
    rating_system = LevesonRatingSystem(storage_file=args.storage)
    
    try:
        if args.command == 'rate':
            if args.exchange not in rating_system.get_all_exchanges():
                rating_system.add_exchange(args.exchange)
            rating_system.rate_exchange(args.exchange)
            
        elif args.command == 'add':
            if args.exchange not in rating_system.get_all_exchanges():
                rating_system.add_exchange(args.exchange)
            rating_system.add_rating(args.exchange, args.criterion, args.rating)
            print(f"Added rating {args.rating} for {args.criterion} to {args.exchange}")
            
        elif args.command == 'view':
            ratings = rating_system.view_ratings(args.exchange)
            print(f"\nRatings for '{args.exchange}':")
            print("=" * 40)
            for criterion, rating in ratings.items():
                if rating is not None:
                    print(f"{criterion.capitalize():12}: {rating:4.2f}")
                else:
                    print(f"{criterion.capitalize():12}: No ratings")
                    
        elif args.command == 'list':
            exchanges = rating_system.get_all_exchanges()
            if exchanges:
                print("Registered exchanges:")
                for exchange in exchanges:
                    ratings = rating_system.view_ratings(exchange)
                    avg = sum(r for r in ratings.values() if r is not None)
                    count = sum(1 for r in ratings.values() if r is not None)
                    overall = avg / count if count > 0 else None
                    if overall:
                        print(f"  {exchange} (avg: {overall:.2f})")
                    else:
                        print(f"  {exchange} (no ratings)")
            else:
                print("No exchanges registered.")
                
        elif args.command == 'report':
            report = rating_system.generate_report()
            print("\nLBTAS System Report")
            print("=" * 50)
            print(f"Total exchanges: {report['total_exchanges']}")
            print(f"Total ratings: {report['total_ratings']}")
            if report['system_average']:
                print(f"System average: {report['system_average']}")
            
            if report['category_averages']:
                print("\nCategory Averages:")
                for category, avg in report['category_averages'].items():
                    if avg:
                        print(f"  {category.capitalize():12}: {avg}")
            
            if report['top_performers']:
                print("\nTop Performers:")
                for exchange, avg in report['top_performers']:
                    print(f"  {exchange}: {avg:.2f}")
                    
        elif args.command == 'export':
            if args.format == 'json':
                with open(args.output, 'w') as f:
                    json.dump(rating_system.exchanges, f, indent=2)
            elif args.format == 'csv':
                rating_system.export_to_csv(args.output)
            print(f"Exported to {args.output}")
            
    except KeyboardInterrupt:
        print("\nOperation cancelled.")
        sys.exit(0)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

# Example usage when run as script
if __name__ == "__main__":
    if len(sys.argv) == 1:
        # Interactive demo mode
        print("LBTAS Interactive Demo")
        print("=" * 50)
        print("Welcome to the Leveson-Based Trade Assessment Scale!")
        print("This demo will walk you through rating a service.\n")
        
        rating_system = LevesonRatingSystem()
        
        try:
            # Demo exchange
            demo_exchange = "DemoService"
            print(f"Creating demo exchange: '{demo_exchange}'")
            rating_system.add_exchange(demo_exchange)
            
            print("\nYou can now rate this service on four criteria:")
            print("- Reliability: How dependable and consistent is it?")
            print("- Usability: How easy and pleasant is it to use?") 
            print("- Performance: How fast and efficient is it?")
            print("- Support: How helpful is customer service?")
            
            print("\nWould you like to rate this demo service? (y/n): ", end="")
            if input().lower().startswith('y'):
                rating_system.rate_exchange(demo_exchange)
                
                # Show results
                ratings = rating_system.view_ratings(demo_exchange)
                print(f"\nYour ratings for '{demo_exchange}':")
                print("-" * 30)
                for criterion, rating in ratings.items():
                    if rating is not None:
                        description = ""
                        if rating >= 4:
                            description = " (Delight!)"
                        elif rating >= 3:
                            description = " (Excellent)"
                        elif rating >= 2:
                            description = " (Good)"
                        elif rating >= 1:
                            description = " (Acceptable)"
                        elif rating >= 0:
                            description = " (Minimal)"
                        else:
                            description = " (Problematic)"
                        
                        print(f"{criterion.capitalize():12}: {rating:4.1f}{description}")
                
                avg_rating = sum(r for r in ratings.values() if r is not None) / len([r for r in ratings.values() if r is not None])
                print(f"\nOverall Average: {avg_rating:.2f}")
            else:
                print("Demo completed. Try 'python lbtas.py --help' for CLI options.")
                
        except KeyboardInterrupt:
            print("\nDemo cancelled. Goodbye!")
        except Exception as e:
            print(f"Demo error: {e}")
    else:
        # Use CLI interface
        main()

File: lbtas.ts
/**
 * Leveson-Based Trade Assessment Scale (LBTAS)
 * 
 * A rating system for digital commerce based on Nancy Leveson's
 * aircraft software assessment methodology.
 * 
 * Copyright (C) 2024 Network Theory Applied Research Institute
 * Licensed under GNU Affero General Public License v3.0
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 */

import * as fs from 'fs';
import * as readline from 'readline';

const VERSION = '1.0.0';
const AUTHOR = 'Network Theory Applied Research Institute';
const LICENSE = 'AGPL-3.0';

interface ExchangeData {
  [category: string]: number[];
  _metadata: {
    created: string;
    total_ratings: number;
  };
}

interface StorageData {
  [exchange: string]: ExchangeData;
}

interface RatingSummary {
  [category: string]: number | null;
}

interface SystemReport {
  total_exchanges: number;
  total_ratings: number;
  system_average: number | null;
  category_averages: { [key: string]: number | null };
  top_performers: Array<[string, number]>;
  bottom_performers: Array<[string, number]>;
}

class LevesonRatingSystem {
  private readonly DEFAULT_CATEGORIES = ['reliability', 'usability', 'performance', 'support'];
  
  private readonly RATING_DESCRIPTIONS: { [key: number]: string } = {
    '-1': 'No Trust - User was harmed, exploited, or received a product or service with no discipline or malicious intent.',
    '0': 'Cynical Satisfaction - Interaction fulfills a basic promise requiring little to no discipline toward user satisfaction.',
    '1': 'Basic Promise - Interaction meets all articulated user demands, no more.',
    '2': 'Basic Satisfaction - Interaction meets socially acceptable standards exceeding articulated user demands.',
    '3': 'No Negative Consequences - Interaction designed to prevent loss, exceed basic quality.',
    '4': 'Delight - Interaction anticipates the evolution of user practices and concerns post-transaction.'
  };

  private categories: string[];
  private storagePath: string | null;
  private exchanges: StorageData;

  constructor(storagePath: string | null = null, categories: string[] | null = null) {
    this.categories = categories || this.DEFAULT_CATEGORIES.slice();
    this.storagePath = storagePath;
    this.exchanges = {};

    if (this.storagePath) {
      this.loadFromFile();
    }
  }

  private loadFromFile(): void {
    if (!this.storagePath) return;
    
    try {
      if (fs.existsSync(this.storagePath)) {
        const data = fs.readFileSync(this.storagePath, 'utf8');
        this.exchanges = JSON.parse(data);
      }
    } catch (error) {
      console.error(`Error loading from file: ${error}`);
    }
  }

  private saveToFile(): void {
    if (!this.storagePath) return;
    
    try {
      fs.writeFileSync(this.storagePath, JSON.stringify(this.exchanges, null, 2));
    } catch (error) {
      console.error(`Error saving to file: ${error}`);
    }
  }

  addExchange(name: string): void {
    if (this.exchanges[name]) {
      throw new Error(`Exchange '${name}' already exists.`);
    }

    const exchangeData: ExchangeData = {
      _metadata: {
        created: new Date().toISOString(),
        total_ratings: 0
      }
    };

    for (const category of this.categories) {
      exchangeData[category] = [];
    }

    this.exchanges[name] = exchangeData;
    this.saveToFile();
  }

  addRating(exchangeName: string, criterion: string, rating: number): void {
    if (!this.exchanges[exchangeName]) {
      throw new Error(`Exchange '${exchangeName}' does not exist.`);
    }

    if (!this.categories.includes(criterion)) {
      throw new Error(`Criterion '${criterion}' not in valid categories: ${this.categories.join(', ')}`);
    }

    if (!Number.isInteger(rating) || rating < -1 || rating > 4) {
      throw new Error(`Rating must be integer between -1 and 4, got ${rating}`);
    }

    this.exchanges[exchangeName][criterion].push(rating);
    this.exchanges[exchangeName]._metadata.total_ratings++;
    this.saveToFile();
  }

  async getRating(criterion: string): Promise<number> {
    console.log(`\nRate ${criterion.charAt(0).toUpperCase() + criterion.slice(1)}:`);
    console.log('='.repeat(50));
    
    for (const [rating, description] of Object.entries(this.RATING_DESCRIPTIONS)) {
      console.log(`${rating.padStart(2)}: ${description}`);
    }
    
    console.log('='.repeat(50));

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });

    return new Promise((resolve) => {
      const askRating = () => {
        rl.question(`Enter your rating for ${criterion.charAt(0).toUpperCase() + criterion.slice(1)} (-1 to 4): `, (answer) => {
          const rating = parseInt(answer);
          if (!isNaN(rating) && rating >= -1 && rating <= 4) {
            rl.close();
            resolve(rating);
          } else {
            console.log('Please enter a rating between -1 and 4.');
            askRating();
          }
        });
      };
      askRating();
    });
  }

  async rateExchange(name: string): Promise<void> {
    if (!this.exchanges[name]) {
      throw new Error(`Exchange '${name}' does not exist.`);
    }

    console.log(`\nRating '${name}' using Leveson-Based Trade Assessment Scale`);
    console.log('='.repeat(60));

    for (const criterion of this.categories) {
      const rating = await this.getRating(criterion);
      this.exchanges[name][criterion].push(rating);
      this.exchanges[name]._metadata.total_ratings++;
    }

    this.saveToFile();
    console.log(`\nRating completed for '${name}'!`);
  }

  viewRatings(name: string): RatingSummary {
    if (!this.exchanges[name]) {
      throw new Error(`Exchange '${name}' does not exist.`);
    }

    const summary: RatingSummary = {};

    for (const criterion of this.categories) {
      const ratings = this.exchanges[name][criterion];
      if (ratings.length > 0) {
        const avg = ratings.reduce((a, b) => a + b, 0) / ratings.length;
        summary[criterion] = Math.round(avg * 100) / 100;
      } else {
        summary[criterion] = null;
      }
    }

    return summary;
  }

  getAllExchanges(): string[] {
    return Object.keys(this.exchanges);
  }

  generateReport(): SystemReport {
    const totalExchanges = Object.keys(this.exchanges).length;

    if (totalExchanges === 0) {
      return {
        total_exchanges: 0,
        total_ratings: 0,
        system_average: null,
        category_averages: {},
        top_performers: [],
        bottom_performers: []
      };
    }

    const allRatings: number[] = [];
    const categoryTotals: { [key: string]: number[] } = {};
    const exchangeAverages: { [key: string]: number } = {};

    for (const category of this.categories) {
      categoryTotals[category] = [];
    }

    for (const [exchangeName, exchangeData] of Object.entries(this.exchanges)) {
      const exchangeRatings: number[] = [];

      for (const category of this.categories) {
        const ratings = exchangeData[category];
        if (ratings.length > 0) {
          const avg = ratings.reduce((a, b) => a + b, 0) / ratings.length;
          exchangeRatings.push(avg);
          categoryTotals[category].push(avg);
          allRatings.push(...ratings);
        }
      }

      if (exchangeRatings.length > 0) {
        exchangeAverages[exchangeName] = 
          exchangeRatings.reduce((a, b) => a + b, 0) / exchangeRatings.length;
      }
    }

    const systemAverage = allRatings.length > 0
      ? allRatings.reduce((a, b) => a + b, 0) / allRatings.length
      : null;

    const categoryAverages: { [key: string]: number | null } = {};
    for (const [category, ratings] of Object.entries(categoryTotals)) {
      categoryAverages[category] = ratings.length > 0
        ? ratings.reduce((a, b) => a + b, 0) / ratings.length
        : null;
    }

    const sortedExchanges = Object.entries(exchangeAverages)
      .sort((a, b) => b[1] - a[1]);

    const topPerformers = sortedExchanges.slice(0, 5);
    const bottomPerformers = sortedExchanges.slice(-5).reverse();

    return {
      total_exchanges: totalExchanges,
      total_ratings: allRatings.length,
      system_average: systemAverage,
      category_averages: categoryAverages,
      top_performers: topPerformers,
      bottom_performers: bottomPerformers
    };
  }

  exportToJSON(outputPath: string): void {
    fs.writeFileSync(outputPath, JSON.stringify(this.exchanges, null, 2));
  }

  exportToCSV(outputPath: string): void {
    const lines: string[] = ['exchange,category,rating,timestamp'];

    for (const [exchangeName, exchangeData] of Object.entries(this.exchanges)) {
      for (const category of this.categories) {
        const ratings = exchangeData[category];
        for (let i = 0; i < ratings.length; i++) {
          lines.push(`${exchangeName},${category},${ratings[i]},${i + 1}`);
        }
      }
    }

    fs.writeFileSync(outputPath, lines.join('\n'));
  }
}

// CLI interface
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.log('LBTAS - Leveson-Based Trade Assessment Scale');
    console.log('\nUsage:');
    console.log('  node lbtas.js rate <exchange>');
    console.log('  node lbtas.js add <exchange> <criterion> <rating>');
    console.log('  node lbtas.js view <exchange>');
    console.log('  node lbtas.js list');
    console.log('  node lbtas.js report');
    console.log('  node lbtas.js export <format> <output>');
    return;
  }

  const command = args[0];
  const storage = 'lbtas_ratings.json';
  const system = new LevesonRatingSystem(storage);

  try {
    switch (command) {
      case 'rate': {
        const exchange = args[1];
        if (!exchange) {
          console.error('Error: Exchange name required');
          return;
        }
        if (!system.getAllExchanges().includes(exchange)) {
          system.addExchange(exchange);
        }
        await system.rateExchange(exchange);
        break;
      }

      case 'add': {
        const [, exchange, criterion, ratingStr] = args;
        if (!exchange || !criterion || !ratingStr) {
          console.error('Error: exchange, criterion, and rating required');
          return;
        }
        const rating = parseInt(ratingStr);
        if (!system.getAllExchanges().includes(exchange)) {
          system.addExchange(exchange);
        }
        system.addRating(exchange, criterion, rating);
        console.log(`Added rating ${rating} for ${criterion} to ${exchange}`);
        break;
      }

      case 'view': {
        const exchange = args[1];
        if (!exchange) {
          console.error('Error: Exchange name required');
          return;
        }
        const ratings = system.viewRatings(exchange);
        console.log(`\nRatings for '${exchange}':`);
        console.log('='.repeat(40));
        for (const [criterion, rating] of Object.entries(ratings)) {
          if (rating !== null) {
            console.log(`${criterion.padEnd(12)}: ${rating.toFixed(2)}`);
          } else {
            console.log(`${criterion.padEnd(12)}: No ratings`);
          }
        }
        break;
      }

      case 'list': {
        const exchanges = system.getAllExchanges();
        if (exchanges.length === 0) {
          console.log('No exchanges registered.');
        } else {
          console.log('Registered exchanges:');
          for (const exchange of exchanges) {
            const ratings = system.viewRatings(exchange);
            const values = Object.values(ratings).filter(r => r !== null) as number[];
            const overall = values.length > 0
              ? values.reduce((a, b) => a + b, 0) / values.length
              : null;
            if (overall !== null) {
              console.log(`  ${exchange} (avg: ${overall.toFixed(2)})`);
            } else {
              console.log(`  ${exchange} (no ratings)`);
            }
          }
        }
        break;
      }

      case 'report': {
        const report = system.generateReport();
        console.log('\nLBTAS System Report');
        console.log('='.repeat(50));
        console.log(`Total exchanges: ${report.total_exchanges}`);
        console.log(`Total ratings: ${report.total_ratings}`);
        if (report.system_average !== null) {
          console.log(`System average: ${report.system_average.toFixed(2)}`);
        }

        if (Object.keys(report.category_averages).length > 0) {
          console.log('\nCategory Averages:');
          for (const [category, avg] of Object.entries(report.category_averages)) {
            if (avg !== null) {
              console.log(`  ${category.padEnd(12)}: ${avg.toFixed(2)}`);
            }
          }
        }

        if (report.top_performers.length > 0) {
          console.log('\nTop Performers:');
          for (const [exchange, avg] of report.top_performers) {
            console.log(`  ${exchange}: ${avg.toFixed(2)}`);
          }
        }
        break;
      }

      case 'export': {
        const format = args[1];
        const output = args[2];
        if (!format || !output) {
          console.error('Error: format and output path required');
          return;
        }
        if (format === 'json') {
          system.exportToJSON(output);
        } else if (format === 'csv') {
          system.exportToCSV(output);
        } else {
          console.error('Error: format must be json or csv');
          return;
        }
        console.log(`Exported to ${output}`);
        break;
      }

      default:
        console.error(`Unknown command: ${command}`);
    }
  } catch (error) {
    console.error(`Error: ${error}`);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

export { LevesonRatingSystem };
